core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMetadata.java

```diff
@@ -306,14 +307,57 @@ default Map<SchemaTableName, List<ColumnMetadata>> listTableColumns(ConnectorSes
     /**
      * Gets the metadata for all columns that match the specified table prefix. Redirected table names are included, but
      * the column metadata for them is not. Views and materialized views are not included.
+     *
+     * @deprecated Implement {@link #streamRelationColumns}.
      */
+    @Deprecated
     default Iterator<TableColumnsMetadata> streamTableColumns(ConnectorSession session, SchemaTablePrefix prefix)
     {
         return listTableColumns(session, prefix).entrySet().stream()
                 .map(entry -> TableColumnsMetadata.forTable(entry.getKey(), entry.getValue()))
                 .iterator();
     }
 
+    /**
+     * Gets columns for all relations (tables, views, materialized views), possibly filtered by schemaName.
+     * (e.g. for all relations that would be returned by {@link #listTables(ConnectorSession, Optional)}).
+     * Redirected table names are included, but the comment for them is not.
+     */
+    @Experimental(eta = "2024-01-01")
+    default Iterator<TableColumnsMetadata> streamRelationColumns(
+            ConnectorSession session,
+            Optional<String> schemaName,
+            UnaryOperator<Set<SchemaTableName>> tablesFilter,
+            BiFunction<SchemaTableName, ConnectorViewDefinition.ViewColumn, ColumnMetadata> viewColumnMetadata,
+            BiFunction<SchemaTableName, ConnectorMaterializedViewDefinition.Column, ColumnMetadata> materializedViewColumnMetadata)
```

Comment:
There are two filtering parameters: `schemaName` and `tabesFilter`. Application of both is mandatory.
Encapsulating them in a class would allow us to introduce a new mandatory filter without implementation knowing about it, which would be a silent breaking change. So no, i don't think it's a good idea.
BTW the reason why `schemaName` filter is mandatory is probably obvious. `tabesFilter` is mandatory for performance reasons. AccessControl checks can be expensive (can involve reading permissions for given relation from a remote system) and should not be done more than once.